server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -298,35 +299,95 @@ public void parse(ParseContext context) throws IOException {
+                List<Object> values = new ArrayList<>(sourceValues.size());
+                for (Object value : sourceValues) {
+                    Object converted = convertSourceValue.apply(value);
+                    if (converted != null) {
+                        values.add(converted);
+                    }
                 }
+                return values;
             }
-        }
-        return values;
+            Object converted = convertSourceValue.apply(sourceValue);
+            if (converted != null) {
+                return List.of(converted);
+            }
+            return List.of();
+        });
     }
 
     /**
-     * Given a value that has been extracted from a document's source, parse it into a standard
-     * format. This parsing logic should closely mirror the value parsing in
-     * {@link #parseCreateField} or {@link #parse}.
-     *
-     * Note that when overriding this method, {@link #lookupValues} should *not* be overridden.
+     * Build a {@linkplain ValueFetcher} that reads values from the source
+     * then calls {@code convertSourceValues} to parse, normalize, and format the value.
+     * <p>
+     * If the source doesn't contain the value then this will return an
+     * empty list and won't call {@code convertSourceValues} at all.
+     */
+    protected final ValueFetcher sourceListValueFetcher(
+        SearchLookup lookup,
+        CheckedFunction<Object, List<?>, IOException> convertSourceValues
+    ) {
+        assert parsesArrayValue();
+        return uncheckedSourceValueFetcher(lookup, convertSourceValues);
+    }
+
+    private ValueFetcher uncheckedSourceValueFetcher(
+        SearchLookup lookup,
+        CheckedFunction<Object, List<?>, IOException> convertSourceValues
+    ) {
+        return ctx -> docId -> {
+            /*
+             * setSegmentAndDocument out of pure paranoia. The fetch phase should already
+             * have positioned us at the right document. So we assert that too.
+             */
+            boolean alreadyPositioned = lookup.source().setSegmentAndDocument(ctx, docId);
+            assert alreadyPositioned;
+            Object sourceValue = lookup.source().extractValue(name(), nullValue());
+            if (sourceValue == null) {
+                return List.of();
+            }
+            return convertSourceValues.apply(sourceValue);
+        };
+    }
+
+    /**
+     * Build a {@linkplain ValueFetcher} that reads values from doc values using the
+     * same logic as the doc values fetch phase.
      */
-    protected abstract Object parseSourceValue(Object value, @Nullable String format);
+    protected final ValueFetcher docValuesFetcher(SearchLookup lookup, String format) {
+        DocValueFormat dvFormat = fieldType().docValueFormat(format, null);
+        IndexFieldData<?> fd = lookup.doc().getForField(fieldType());
+        return ctx -> fd.load(ctx).buildFetcher(dvFormat);
+    }
```

Comment:
I decided to implement fetching runtime fields by pointing them at the doc values. This gave me a convenient excuse to remove most of the `instanceof`s in `FetchDocValuesPhase` while implementing my feature. As a bonus all of the tests for fetching doc values cover this.

Reply:
I do imagine a day where we can "fake out" doc values from `_source`, but I think we'd do that by building a "funny SearchLookup".